madminer/utils/morphing.py

```diff
@@ -396,33 +604,61 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
             raise RuntimeError(
                 "No components defined. Use morpher.set_components() or morpher.find_components() " "first!"
             )
+        if self.gd is None and self.gp is None and self.gs is None:
+            if basis is None:
+                basis = self.basis
+                morphing_matrix = self.morphing_matrix
+
+            if basis is None:
+                raise RuntimeError(
+                    "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
+                    "basis keyword."
+                )
 
-        if basis is None:
-            basis = self.basis
-            morphing_matrix = self.morphing_matrix
+            if morphing_matrix is None:
+                morphing_matrix = self.calculate_morphing_matrix(basis)
 
-        if basis is None:
-            raise RuntimeError(
-                "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
-                "basis keyword."
-            )
+            # Calculate component weights
+            component_weights = np.zeros(self.n_components)
+            for c in range(self.n_components):
+                factor = 1.0
+                for p in range(self.n_parameters):
+                    factor *= float(theta[p] ** self.components[c, p])
+                component_weights[c] = factor
```

Comment:
This is occurrence number 1 of the _component weights calculation_ snippet.

Please extract these lines to a private method within this class.

Reply:
Sure, I'll extract these to a private method